core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicatedMutationAction.java

```diff
@@ -0,0 +1,162 @@
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.index.translog.Translog;
+import org.elasticsearch.index.translog.Translog.Location;
+import org.elasticsearch.indices.IndicesService;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportResponse;
+import org.elasticsearch.transport.TransportService;
+
+import java.util.function.Supplier;
+
+/**
+ * Base class for transport actions that modify data in some shard like index, delete, and shardBulk.
+ */
+public abstract class TransportReplicatedMutationAction<
+            Request extends ReplicatedMutationRequest<Request>,
+            Response extends ReplicatedMutationResponse
+        > extends TransportReplicationAction<Request, Request, Response> {
+
+    protected TransportReplicatedMutationAction(Settings settings, String actionName, TransportService transportService,
+            ClusterService clusterService, IndicesService indicesService, ThreadPool threadPool, ShardStateAction shardStateAction,
+            ActionFilters actionFilters, IndexNameExpressionResolver indexNameExpressionResolver, Supplier<Request> request,
+            String executor) {
+        super(settings, actionName, transportService, clusterService, indicesService, threadPool, shardStateAction, actionFilters,
+                indexNameExpressionResolver, request, request, executor);
+    }
+
+    /**
+     * Called with a reference to the primary shard.
+     *
+     * @return the result of the write - basically just the response to send back and the translog location of the {@linkplain IndexShard}
+     *         after the write was completed
+     */
+    protected abstract WriteResult<Response> onPrimaryShard(IndexService indexService, IndexShard indexShard, Request request)
+            throws Exception;
+
+    /**
+     * Called once per replica with a reference to the {@linkplain IndexShard} to modify.
+     *
+     * @return the translog location of the {@linkplain IndexShard} after the write was completed
+     */
+    protected abstract Translog.Location onReplicaShard(Request request, IndexShard indexShard);
+
+    @Override
+    protected Request shardOperationOnPrimary(Request request, ActionListener<Response> listener) throws Exception {
+        IndexService indexService = indicesService.indexServiceSafe(request.shardId().getIndex());
+        IndexShard indexShard = indexService.getShard(request.shardId().id());
+        WriteResult<Response> result = onPrimaryShard(indexService, indexShard, request);
+        processAfterWrite(request.refresh(), indexShard, result.location);
+        if (request.refresh()) {
+            // Only setForcedRefresh if it is true because this can touch every item in a bulk request
+            result.response.setForcedRefresh(true);
+        }
+        if (request.shouldBlockUntilRefresh() && false == request.refresh()) {
+            indexShard.addRefreshListener(new RefreshListener() {
```

Comment:
I think we should make this refresh listener a static inner class just to make sure we are not holding on to unnecessary memory?